### ISSUE TITLE
Ghost window with KDE Connect 

### ISSUE THREAD

**Issue by MithicSpirit** [qtile/qtile#3308]
How to reproduce:
1. Install/start/configure KDE Connect
2. Open "Slideshow remote" (on phone)
3. Enable the pointer (tap & hold)
4. Besides the overlay window used to draw the pointer (which is floating), there appears to be another "ghost" window allocated in the layout; disabling the pointer makes it go away
5. Enabling the pointer again, move the overlay window (super+drag) so it remains floating but is no longer maximized; the ghost window remains
6. Disabling the pointer now makes ghost window stay there despite overlay window being gone
7. Restarting qtile (e.g. `qtile cmd-obj -o cmd -f restart`) makes the ghost window go away again

Note: this seems to happen on all layouts but it is easiest to observe in the `monadtall` layout with one other window open.

The expected behavior is that the ghost window never exist in the first place (step 4), although it is also unexpected that it should linger behind after disabling the pointer (step 6).

Additional info:
- `qtile --version`: `0.20.0`
- `python --version`: `Python 3.10.2`
- OS: Arch Linux
- Window system: X.org (with `picom` as compositor)

**Comment by m-col** [comment]
If you `xprop` on the ghost window does it give you information about anything other than the root?

Could you get the window ID from `qtile cmd-obj -o cmd -f windows` and then get its info from `qtile cmd-obj -o window $WID if info`, before and after you disable the pointer?

Does `wmctrl -l` give you any information about it?

**Comment by MithicSpirit** [comment]
`xprop` just gives info about the root window when taken on the ghost window, but it properly shows the information of the overlay (icons omitted as no meaningful information could be shown): 
```
_VARIABLE_REFRESH(CARDINAL) = 1
_NET_FRAME_EXTENTS(CARDINAL) = 0, 0, 0, 0
_NET_WM_DESKTOP(CARDINAL) = 5
WM_STATE(WM_STATE):
		window state: Normal
		icon window: 0x0
_NET_WM_USER_TIME(CARDINAL) = 3809284
_NET_WM_STATE(ATOM) = _NET_WM_STATE_ABOVE, _NET_WM_STATE_FULLSCREEN, _NET_WM_STATE_STAYS_ON_TOP
_NET_WM_ICON(CARDINAL) =
XdndAware(ATOM) = BITMAP
WM_NAME(STRING) =
_NET_WM_NAME(UTF8_STRING) = "KDE Connect Daemon"
_MOTIF_WM_HINTS(_MOTIF_WM_HINTS) = 0x2, 0x1, 0x0, 0x0, 0x0
_NET_WM_WINDOW_TYPE(ATOM) = _KDE_NET_WM_WINDOW_TYPE_OVERRIDE, _NET_WM_WINDOW_TYPE_NORMAL
_XEMBED_INFO(_XEMBED_INFO) = 0x0, 0x1
WM_CLIENT_LEADER(WINDOW): window id # 0x6600015
WM_HINTS(WM_HINTS):
		Client accepts input or input focus: False
		Initial state is Normal State.
		window id # of group leader: 0x6600015
WM_CLIENT_MACHINE(STRING) = "archimedes"
_NET_WM_PID(CARDINAL) = 3785
_NET_WM_SYNC_REQUEST_COUNTER(CARDINAL) = 106954881
WM_CLASS(STRING) = "kdeconnectd", "kdeconnect.daemon"
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, WM_TAKE_FOCUS, _NET_WM_PING, _NET_WM_SYNC_REQUEST
WM_NORMAL_HINTS(WM_SIZE_HINTS):
		user specified size: 0 by 0
		window gravity: Static
```

`qtile cmd-obj -o cmd -f windows` doesn't work for me and just raises an `AttributeError`:
```
$ qtile cmd-obj -o cmd -f windows                                                         11:40:35
error: Sorry cannot run function 'windows' with arguments []: Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/libqtile/command/interface.py", line 305, in call
    return SUCCESS, cmd(*args, **kwargs)
  File "/usr/lib/python3.10/site-packages/libqtile/core/manager.py", line 1230, in cmd_windows
    return [i.info() for i in self.windows_map.values() if not isinstance(i, base.Internal)]
  File "/usr/lib/python3.10/site-packages/libqtile/core/manager.py", line 1230, in <listcomp>
    return [i.info() for i in self.windows_map.values() if not isinstance(i, base.Internal)]
  File "/usr/lib/python3.10/site-packages/libqtile/backend/x11/window.py", line 684, in info
    wm_class=self.get_wm_class(),
  File "/usr/lib/python3.10/site-packages/libqtile/backend/x11/window.py", line 598, in get_wm_class
    return self._wm_class
AttributeError: 'Icon' object has no attribute '_wm_class'
```

`wmctrl -l` doesn't show the ghost window, just the floating/maximized overlay window (whose name is `KDE Connect Daemon`).

Another thing I've noticed is that if I take the overlay window out of being maximized and floating it takes up the spot of the ghost window, so maybe qtile is getting confused on whether it should allocate a spot in the tiling to that window. This is also supported by running `qtile cmd-obj -o groups -f info` during and after the pointer:
- During, the window with name `KDE Connect Daemon` shows up in both `'windows'` and `'tiled_windows'` (it also shows up in `wmctrl -l`)
- After it only shows up in `'tiled_windows'` (and not in `wmctrl -l`)

**Comment by m-col** [comment]
OK I think I figured it out. The overlay window _is_ the "ghost" window. It just looks weird because it's completely transparent. If you fullscreen a regular opaque window, that's what your layout looks like underneath, because fullscreen windows keep their place in the layout so that when they are tiled again the order of windows persists. You just don't see it usually because of the fullscreen window!

Basically you just need to auto-float the overlay window to stop it from taking a slot in the layout.

**Comment by MithicSpirit** [comment]
Still doesn't seem to work. If I add `Match(title="KDE Connect Daemon")` to my `float_rules` (then run `qtile cmd-obj -o cmd -f restart`) the behavior doesn't change.

**Comment by m-col** [comment]
> I suspect that the issue is that it is spawned as a fullscreen window, which means that qtile immediately allocates a spot in the tiling for it, before checking if it should be floating. 

Yeah it looks like this. Here is a working solution though:

```py
@hook.subscribe.client_new
def _kde_connect(win):
    if win.name == "KDE Connect Daemon":
        s = qtile.current_screen
        win.cmd_static(qtile.screens.index(s), s.x, s.y, s.width, s.height)
```

> There is also a second issue going on here where a fullscreen window that is made to float by holding super and clicking+dragging with the mouse will keep its tiling spot allocated, even though floating windows shouldn't have it, and when the window dies, since it was floating, the spot doesn't get deallocated.

Hmm that is definitely a bug. Pinging @jwijenbergh have you noticed this since your change? :smile:

**Comment by MithicSpirit** [comment]
That function/hook fixed it, thank you.

As for that other bug I can open a new issue if you would like. I can consistently reproduce it by taking a window, putting it in maximize, and moving it into floating (super+mouse1 drag).

**Comment by jwijenbergh** [comment]
> As for that other bug I can open a new issue if you would like. I can consistently reproduce it by taking a window, putting it in maximize, and moving it into floating (super+mouse1 drag).

Does that only happen when setting the window to floating through the mouse? What happens when you use a keybind to float and then remove the window?

**Comment by MithicSpirit** [comment]
From my previous testing, using the keybind for floating is a bit tough since `lazy.window.toggle_floating()` just disables it being fullscreen and puts it back in the tiling (likely due to qtile fullscreening windows by making them float). 

I could try making a function to set `qtile.current_window.floating = False`, though I suspect this will cause the same behavior; I think that setting `qtile.current_window.fullscreen = False` might work instead. Sadly I'm not available to test these right now, but I should be in a couple hours and I'll get back to you then. Please let me know if you have any other ideas.

**Comment by MithicSpirit** [comment]
Just tested both and neither of them worked (both just made it tile again), so the only way to reproduce it is using the mouse. Another thing I've noticed is that if one disables/toggles floating on a window that is fullscreen it will unfullscreen correctly but, in `xprop`, it will keep `_NET_WM_STATE_FULLSCREEN`, which can cause odd behavior with things like compositors (e.g. I have rounded corners in my `picom` config, but they are not rounded for that window since it thinks it should be fullscreen).

I believe that both of these problems could be solved by adding some form of properly handling of the floating state on windows that are maximized.

**Comment by jwijenbergh** [comment]
Thanks for the testing! I will look into it

**Comment by jwijenbergh** [comment]
Could you see if #3324 fixes the fullscreen issue? Also does `_NET_WM_STATE` now correctly update?

**Comment by MithicSpirit** [comment]
This has fixed the fullscreen → floating issue, but not `_NET_WM_STATE`.

**Comment by jwijenbergh** [comment]
Done can you test the update on #3324?

**Comment by MithicSpirit** [comment]
Yeah this issue report has really branched out into a lot of issues, sorry about that. I'm a bit busy right now so it might take a bit until I can test the new fix, but I'll try to get back to you on that as soon as I can.

**Comment by MithicSpirit** [comment]
Yup, now with commit d55c4390e95850bf66c0333e1773444b3539e97d the `_NET_WM_STATE` problem has also been fixed. Thank you for your help!

**Comment by jwijenbergh** [comment]
Is this issue completely fixed on master?

**Comment by m-col** [comment]
Yep, thanks!
